### DISCUSSION TITLE
Scenario Community High Score\Leader Board?

### DISCUSSION THREAD

**Discussion by ageekhere** [OpenRCT2/OpenRCT2#17549] [Feature requests]
When you complete a scenario you are given a local high score. I wonder if it would be possible to also have a high score\leader board for each scenario showing the world records. Players would have an option to submit their high score to the leader board. once they complete a mission.

A few challenges would be
1. Prevent scores where cheats were used
2. Prevent hacked scores or hacked scenarios
3. Prevent scores which used a game breaking bug/ exploit
4. Make sure maintaining the high score leader board is not too time consuming or difficult   

Maybe a replay feature could be used to validate scores.

It would be a fun and interesting to see what players will come up with to try and achieve the highest score.

**Comment by ocalhoun6** [comment]
In an open-source game, you'd never be able to fully prevent cheating. There's nothing stopping anybody with the right skills from downloading the source code, modifying it to remove the anti-cheat features, building it, and running it. Then they could use cheats to their heart's content and still upload them to the leaderboard. (Unless there's some kind of server-side checking of the player's code/checksum? But that sounds slow and resource intensive ... and perhaps prone to bugs that might prevent legitimate scores. Oh, and you'd probably have to check/checksum all their object files too. It might be possible to do exploits by modifying objects.)

Will people actually do that? Who knows. But if anybody did, it would quickly ruin the leaderboard feature by having all the top spots filled by cheaters.

There's also a problem number 5: if the leaderboard includes some kind of username/initials along with the score, will there be any sort of moderation to control people adding obscene/offensive or spammy names?

It also opens a can of worms a bit when it comes to "What exactly counts as a cheat?" A lot of OpenRCT2's default quality of life features could be considered 'cheats' from a certain point of view, especially by someone min-maxing every click and keystroke to get the highest score possible. For example, is fast-forwarding a 'cheat'? What about pausing the game a lot? And then there's plugins ... which I guess you'd have to just count all plugins as cheats, just to be safe, even though a lot of plugins wouldn't provide any actual cheat benefit.

Oh ... and problem number 6: while a lot of scenarios have a time limit baked in ... some don't. In, say, a 'complete 5 coasters' scenario, it would be trivial to _almost_ finish the scenario goal, then ignore the scenario goal in favor of driving the park value to an arbitrarily high number, then finish the final coaster when you've hit a satisfactory score. Leaderboards on those types of scenarios could get very crazy, with no upper limit to the possible score, aside from the player's patience ... especially if fast-forwarding is allowed.

**Reply by ageekhere** [reply]
Yes, while it would be a fun idea I think it would be too hard to prevent cheating.

**Reply by IntelOrca** [reply]
Validation is possible through replays, but as you have mentioned... plugins or just changes to the source code would allow the game to be programatically controlled rather than the player. A bit like a tool-assisted speedrun (TAS).

However, I don't know how much is about strategy rather than speed. I assume the former, but strategies are more easily reproduced by other players.

**Reply by ocalhoun6** [reply]
For scenarios with a time limit (eg 'number of guests in park by october year 4') both strategy and speed would be important.

For scenarios without a time limit (eg 'build 10 coasters'), it would be 100% strategy and patience.

(And when it comes down to it, is park value even the best high score to track? Maybe 'time to scenario completion' would be a better high score to track, especially if you allow the early scenario completion option for scenarios with a time limit. That would be a more valid and more difficult to min-max metric to measure performance by, and it would work equally well on pretty much any scenario, except ones with a 'have fun' objective. Though it does kind of turn your high score board into a speedrunning competition.)
